JustSaying.IntegrationTests/Fluent/IntegrationTestBase.cs

```diff
@@ -80,7 +80,17 @@ protected async Task WhenAsync(IServiceCollection services, Func<IMessagePublish
 
             using (var source = new CancellationTokenSource(Timeout))
             {
-                await Task.WhenAny(action(publisher, listener, source.Token), Task.Delay(Timeout, source.Token)).ConfigureAwait(false);
+                var delayTask = Task.Delay(Timeout, source.Token);
```

Comment:
I'm just confused about the need for a delay task if that timeout is already encapsulated by the cancellation token?

Reply:
You can't await a cancellation token, only a task

Reply:
I think Shayne means why can't we trust action to respect the token. It could be blocked on something, it may not be checking the token as frequently as it should etc...

I hadn't realised we were using the same token for the timeout and passing it into the action. That might explain the weirdness when cancelling, my bad. If it was just for the timeout I think it would be ok.

Reply:
Why even have a delay task at all though? The cancellation token source should handle the timeout automatically, not so?

Reply:
Ok, the fact that the same token also goes into the action would explain why cancelling it cancels running tasks and fails tests.
However I think it's the right thing to do, the alternative would be to give `CancellationToken.None` which never expires.

Reply:
@shaynevanasperen my understanding is it's a defensive mistrust of the code under test. It means we fail fast.

Reply:
In answer to the question "why do we wait for two task, one of which is a fixed time delay", we' don't trust CancellationToken to work in all dodgy code  / failure cases.